### DISCUSSION TITLE
Tabs with long content

### DISCUSSION THREAD

**Discussion by yceballost** [Telefonica/mistica-design#542] [Proposals]
Desde movistar españa se necesita usar el componente de tabs con 8 elementos. Aunque el componente lo admite, desde diseño recomendamos no tener tabs con más de 6 elementos.

La definición de las specs diverge de la definición que existe en desktop ya que no existe la elipsis sino que las tabs van a multilinea

- Ajustar specs o componente al comportamiento deseado
- [ ] Review tabs specs

![image](https://user-images.githubusercontent.com/6722153/176172402-a3f474be-5fc3-48ad-bc05-129e5d07064b.png)
https://ibit.ly/hcUF

**Comment by aweell** [comment]
## Truncado de textos

Habría que tener cuidado en truncar texto en elementos base de navegación, esta información es necesaria para el usuario en su totalidad y provoca que este elemento no sea accesible. Mi propuesta aquí sería que los textos hicieran crecer cada una de las tabs proporcionalmente a su contenido sin restricción.

## Overflow en el contenedor

Para casos en los que la cantidad de tabs necesarias sea mayor que la cantidad óptima, una solución podría ser agrupar todos aquellos elementos que no entran en un overflow menú, mostrando los tres puntos y desplegando un dialog con la lista de opciones.

Ej. Github:

![Screenshot 2022-06-28 at 14 06 51](https://user-images.githubusercontent.com/44420072/176174499-8ad10444-f53c-494b-acb3-d403696969e2.png)

_En este ejemplo las tabs no ocupan todo el ancho del contenedor, por tanto se muestran todas_

![Screenshot 2022-06-28 at 14 07 09](https://user-images.githubusercontent.com/44420072/176174692-893657a3-47be-4974-9b3b-00547638fcfb.png)

_Las tabs ocupan más espacio del disponible, se muestra en menú con las que no pueden mostrarse_


## Scrollable arrows

Otra solución es incluir scroll horizontal en el contenedor y permitir el movimiento a través de botones a izquierda y derecha (en función de la posición del scroll)

Ej. Carbon

![Screenshot 2022-06-28 at 14 20 27](https://user-images.githubusercontent.com/44420072/176176995-e0a0c7be-bdf8-48dd-b1a7-3d319b1adddb.png)

https://carbondesignsystem.com/components/tabs/usage#scrollable

**Reply by dsys-col** [reply]
En mobile este componente funciona de forma parecida, haciendo un scroll horizontal. Hasta ahora es algo que está funcionando con éxito. Creo que la última solución es la más adecuada y parecida a lo que ocurre en dispositivos pequeños. El overflow es algo extraño y demasiado diferente a lo que ocurre en mobile.

**Reply by yceballost** [reply]
que se parezca a mobile para mí no es un argumento válido porque el uso es muy diferente en un entorno desktop. No es lo mismo interactuar con un dedo que con un cursor

**Reply by dsys-col** [reply]
Ese tipo de pensamiento es un poco anticuado y hay que empezar a desterrarlo. Hay muchas pantallas de escritorio que son táctiles ya. Por otra parte, no veo el problema con el cursor.

**Reply by dsys-col** [reply]
Sinceramente, no veo que la primera sea mejor que la segunda. Está claro que ninguna de las dos es lo ideal y que ambas tienen sus pros y contras, pero tampoco veo más alternativas ...salvo que se nos ocurra otra cosa mejor. Habrá que tomar una decisión.

**Reply by yceballost** [reply]
Cómo definirias la interacción del segundo caso? cuando le doy a la flecha cuánto se desplaza el scroll? Estaba pensando que esa opción igual era más "transparente" para el usuario porque el contenido quedaría menos oculto, pero pensándolo bien realmente no es así porque vas a necesitar más clicks para poder ver que otras cosas existen, con la primera opción es cierto que ocultas mucho más el contenido pero con un click puedes ver el resto de contenido.

como dices, ambas opciones tienen sus cosas que no son del todo optimas, tb es cierto que estamos solucionando algo que debería evitarse y por algo será 😁

a nivel técnico si que veo menos problemas en la primera opción

**Reply by aweell** [reply]
Aquí en lo referente a la interacción por teclado el overflow también simplifica su uso, no sería necesario dividir dos niveles de foco (Uno para las tabs y otro para las flechas de scroll), recorro los elementos uno a uno y cuando llego al último soy capaz, sin abandonar el flujo de tabeo, de continuar a los siguientes elementos. 

Aunque se produzca scroll automático a medida que me muevo entre tabs, debería pasar posteriormente por la flecha si quiero entrar dentro del contenido de una determinada página, un paso innecesario. 

(Aquí hay que contar con que deberíamos arreglar el componente para que el movimiento entre tabs sea a través de las flechas en función de su orientación y no con tabulador [Salvo que se planteen tabs de navegación con un link por debajo])

Por poner algún ejemplo más en la palestra :)

Microsoft fluent plantea la misma solución que GitHub:
https://developer.microsoft.com/en-us/fluentui#/controls/web/pivot

Salesforce lightning se suma al overflow menu:
https://www.lightningdesignsystem.com/components/tabs/

**Reply by dsys-col** [reply]
Entiendo lo que me comentáis, pero yo particularmente sigo inclinandome más por los chevrones. Tenemos un ejemplo muy claro del funcionamiento en nuestro menú. En este caso no utilizamos un overflow. Está funcionando muy bien y sin problemas con los chevrones. Si ahí le decimos que el comportamiento es con chevron, no veo la razón por la que no hacerlo en un tab. Y si estamos reproduciéndo este comportamiento en el menú principal, quizás sea buena idea hacer lo mismo en los tabs que desborden y no tener un comportamientos diferentes, ni por dispositivos ni por componente. Cuando pulsas el chevron, caben los siguientes items ocultos en una animación de desplazamiento de derecha a izquierda. No vas de uno en uno y por tanto no necesitas más clics.

<img width="501" alt="menu" src="https://user-images.githubusercontent.com/81299625/176882045-9ba03980-00d9-4b2a-8b0c-71217d7fdaf7.png">
<img width="492" alt="menu2" src="https://user-images.githubusercontent.com/81299625/176882049-bde919ff-cde2-47d6-9150-773f1e9ec307.png">
<img width="486" alt="menu3" src="https://user-images.githubusercontent.com/81299625/176882052-c27821e5-3008-4a3a-8e3b-23ea511482ba.png">

**Reply by yceballost** [reply]
para mí esto no es el componente tab, al menos como está definido en Mística. Para mi esto que muestras es un carrusel
https://ibit.ly/6P0f

**Reply by dsys-col** [reply]
No, hombre, es un ejemplo. Ya sé que no es el componente tab. Sin entrar en otro tipo de valoraciones, esto se hizo mucho antes de tener siquiera ese componente creado. Es solamente para que se vea que la solución con chevrones funciona y ya lo tenemos así en el site. No me imagino esto con overflow. No funciona igual.

**Reply by mariaestevezgarciatelefonica** [reply]
Hola! Participo en este foro por primera vez así que si hay algo en lo que fallo let me know :)

Haciendo un poco de bench y pensando en la mejor experiencia posible, en el caso del overflow veo patrones repetidos como en el caso de github.

¿El ejemplo más reconocido? Microsoft word/google docs...
<img width="1440" alt="Captura de pantalla 2022-07-27 a las 13 43 35" src="https://user-images.githubusercontent.com/110095203/181238728-4b08343a-f8f2-4da4-b121-9da4b6b696d5.png">

Otro caso donde repite es en el sistema de diseño Polaris para Shopify
<img width="655" alt="Captura de pantalla 2022-07-27 a las 13 46 28" src="https://user-images.githubusercontent.com/110095203/181239183-0c295827-9cbd-44aa-9e17-fa13f245aebb.png">


Igualmente apoyo la moción de hacer un test A/B.

Lanzo otra pregunta menos enfocado al continente y más al contenido... ¿Cuánto de necesario es tener más de n tabs? ¿Quizás tiene sentido re-agrupar contenido para no forzar la carga cognitiva del usuario?

**Comment by aweell** [comment]
Por entender un poco el contexto en el que se coloca el componente, son unas tabs que segregan un contenido que forma parte de una página o son tabs que cambian el contenido de toda la página?

**Reply by dsys-col** [reply]
Cambia el contenido que se visualiza justo debajo de cada tab. No cambia la página al completo.

**Comment by yceballost** [comment]
Debido a la división de opiniones que hay y que no hay una clara diferencia entre ambas de cuál puede funcionar mejor, podríamos plantear un test de usuario para valorar qué opción es la más óptima y que decida el poder del research ✨

**Reply by yceballost** [reply]
Patricia Ramirez propone hacer un click test con tareas para evaluar con cuál se entiende mejor el usuario, randomizando los ejercicios para que el usuario no pueda asociar atajos.

**Reply by dsys-col** [reply]
Me parece una idea muy razonable. Estaba a punto de proponer lo mismo.

**Comment by aweell** [comment]
Añado por aquí otro caso en mobile:

![Screenshot 2023-03-27 at 18 39 42](https://user-images.githubusercontent.com/44420072/228007457-5910d75f-bf0f-4e2f-8397-be841651c4c4.png)

Cuando el contenido es excesivamente largo o el usuario ha modificado el tamaño de texto de su dispositivo, aunque el número de tabs es pequeño, la definición actual provoca que el contenido se salga de su contenedor.

Dejo un Figma con posibles soluciones:

https://www.figma.com/file/dhGTgSB5t7s0IGnkj6vjRA/Tabs-long-content-(mobile)?node-id=0%3A1&t=YxSpftknvh18oWSw-1

**Reply by jpadronl** [reply]
Se ha transformado esto en algún tipo de ticket JIRA? Cual es el siguiente paso?

**Reply by yceballost** [reply]
yo tiraría la por la opción 1 en cuanto a lo que mística implica. Si el equipo de diseño quiere considerar la opción 3 pues es toda su decisión.

@jpadronl genero issue de esto, actualizamos specs y crearemos ticket en jira a webcore

#1161

**Comment by AnaMontes11** [comment]
Cierro discussion. Nos ponemos con la opción 1 en la tarea #1161

**Comment by MontseTelefonica** [comment]
Hola compañeros! Justo se nos da este problemas en las tabs del CMS de modular pages. Al aumentar el numero de tabs el contenedor se amplia y se nos recorta la preview, pero si ponemos un ancho al contenedor de las tabs no tenemos forma de hacer scroll a los elementos ocultos. Necesitamos desbloquear estol
<img width="1416" alt="image" src="https://github.com/Telefonica/mistica-design/assets/95479574/b32ff596-9f22-4cd0-bdc2-d422fe082ea6">

**Reply by yceballost** [reply]
El componente tab necesita un rediseño fuerte en desktop, no termina de ajustarse bien en muchos de los casos y es bastante rigido.

Has pensado una opción pasajera de usar Chips como selectores?

[Ejemplo](https://mistica-web.vercel.app/playroom/preview/#?code=N4Igxg9gJgpiBcIA8AlGBnADhAduglgG4wAyAhgJ4QCuALgHwA6OABC0gEIQAeLmZUKPhwBzAJoBeYACYALAF8mrNuwDKtMmADWLLJphTZABkXMVK1APwQA4gCcamFjjIBbA43AALfJgC0Ig7UmJ4ssABmZNQANrQAamTR1B4gAIyeSubmSACSONHCMLr8YAbAABymylkWKFYQHHS0uGY1WYSJyRKe6SCtbSp2MDiwdlIAFMAsYF4w2jBQADQs0WQARjDROVAs8gCULBL0LOP9AxYAwj5OmrREZTNzWgvyLPhQUqsbW1BV520XXC0Ya0M7-JAAeiuvky-xYe3kYKyENhA0sQgaTRa1XOHSSKWkniRNSGIxgY2Ak2ms3mSxW602212ByOJ2JbSQ0JuYDuxCkj1pr3enwZPz+cPMAEFqEJhqV2TVIVzUecEQqWCj1ejrI1aM0cOq2Hiup4AMxEnEDUmjCZTAXPOlfRk7faHY6nS3nTnXFi3e78mkOoUfYBOsUqiWSsCldAENb4Aq0CiGixQ64Rtpqz3mTXZ2r1XX6lPGlKyC0S63k23Up4LZZhpmu1keiVsb2+X08-3Ae0vN4hhu-DP-AAqdk0WmEIhTbbTMJTWf+ufBdQxhexEpL3RAAFZy3DKxSqb3HaLGyz3TP2FzO7yHoG+8LQ2eh1e2ABFSVXpXpheIvNsMuXqrjqWIGgBKhbp4ABs+7-Ie1YnvWL7Mm6bIQdkN5+nyPYPr8-Yit82ziq2LAACIAKJxN+c6YMOWSLucQGKhCeQFDgMAZpCIG2EEdFnJC6gTrCkJcNwSjcRg2B4Pc5BUHQSggIsIC0LM7gICAACyECEPg6AaHYIDyEAA)
